src/compiler/checker.ts

```diff
@@ -38599,13 +38607,52 @@ namespace ts {
                     error(node, message, symbolToString(symbol));
                 }
 
-                // Don't allow to re-export something with no value side when `--isolatedModules` is set.
+                // Type assertion to defeat the type predicate aliasing of `node` here because
+                // `isTypeOnlyImportOrExportDeclaration` is a one-sided type guard so `!isDeclaredTypeOnly`
+                // does not actually imply any narrowing effect on `node`.
+                const isDeclaredTypeOnly = isTypeOnlyImportOrExportDeclaration(node as never);

```

Comment:
Would casting `node` to the wider type work instead? Casting to `never` could result in us missing a case in a different change sometime in the future.

Also, we could instead modify `TypeOnlyCompatibleAliasDeclaration` to be a stricter and more accurate definition like:

```ts
export type TypeOnlyCompatibleAliasDeclaration =
    | ImportClause & { readonly isTypeOnly: true }
    | ImportEqualsDeclaration & { readonly isTypeOnly: true }
    | NamespaceImport & { readonly parent: ImportClause & { readonly isTypeOnly: true } }
    | ImportSpecifier & { readonly parent: NamedImports & { readonly parent: ImportClause & { readonly isTypeOnly: true } } }
    | ExportSpecifier & { readonly parent: NamedExports & { readonly parent: ExportDeclaration & { readonly isTypeOnly: true } } }
    ;
```